src/odemis/acq/stitching/_tiledacq.py

```diff
@@ -586,40 +620,73 @@ def _acquireTiles(self):
+        if self._focusing_method in (FocusingMethod.NONE, FocusingMethod.MAX_INTENSITY_PROJECTION):
             return das
-        if i == 0:
-            # Use initial optical focus level to be compared to next tiles
-            # TODO: instead of using the first image, use the best 10% images (excluding outliers)
-            self._good_focus_level = current_focus_level
-
-        # TODO: handle the case of _good_focus_level == 0
-        logging.debug("Current focus level: %s (good = %s)", current_focus_level, self._good_focus_level)
-        # Run autofocus if current focus got worse than permitted deviation
-        if abs(current_focus_level - self._good_focus_level) / self._good_focus_level > FOCUS_FIDELITY:
+        elif self._focusing_method == FocusingMethod.ON_LOW_FOCUS_LEVEL:
+            if i % SKIP_TILES != 0:
+                logging.debug("Skipping focus adjustment..")
+                return das
+
             try:
-                self._future.running_subf = AutoFocus(self._focus_stream.detector,
-                                                      self._focus_stream.emitter,
-                                                      self._focus_stream.focuser,
-                                                      good_focus=self._good_focus,
-                                                      rng_focus=self._focus_rng,
-                                                      method=MTD_EXHAUSTIVE)
-                self._future.running_subf.result()  # blocks until autofocus is finished
-                if self._future._task_state == CANCELLED:
-                    raise CancelledError()
-            except CancelledError:
-                raise
-            except Exception as ex:
-                logging.exception("Running autofocus failed on image i= %s." % i)
-            else:
-                # Reacquire the out of focus tile (which should be corrected now)
-                das = self._getTileDAs(i, ix, iy)
+                current_focus_level = MeasureOpticalFocus(das[self._streams.index(self._focus_stream)])
+            except IndexError:
+                logging.warning("Failed to get image to measure focus on.")
+                return das
+
+            if i == 0:
+                # Use initial optical focus level to be compared to next tiles
+                # TODO: instead of using the first image, use the best 10% images (excluding outliers)
+                self._good_focus_level = current_focus_level
+                return das
+
+            logging.debug("Current focus level: %s (good = %s)", current_focus_level, self._good_focus_level)
+
+            # Run autofocus if current focus got worse than permitted deviation,
+            # or it was very bad originally.
+            if not (self._good_focus_level == 0 or
+                    (self._good_focus_level - current_focus_level) / self._good_focus_level > FOCUS_FIDELITY
+                   ):
```

Comment:
Maybe this can be easier as: 
```
if ( self._good_focus_level != 0
and (self._good_focus_level - current_focus_level)  self._good_focus_level
                <= FOCUS_FIDELITY):
```